cmd/containerd-shim-runhcs-v1/service.go

```diff
@@ -460,7 +455,7 @@ func (s *service) DiagPid(ctx context.Context, req *shimdiag.PidRequest) (*shimd
 	if s == nil {
 		return nil, nil
 	}
-	ctx, span := trace.StartSpan(ctx, "DiagPid")
+	ctx, span := trace.StartSpan(ctx, "DiagPid") //nolint:ineffassign,staticcheck
```

Comment:
It complained because the call to `StartSpan` returns a new ctx, which isn't used again in the function.

Reply:
Ah!

would something like this work for those as well?

```suggestion
	_, span := trace.StartSpan(ctx, "DiagPid")
```

Reply:
Can we do `_, span := ...` and remove the `nolint`?

Reply:
Oh, @thaJeztah commented right before me :)

Reply:
🙈sorry 😂

Reply:
I did this just because I think in the future if someone is adding code later in the function to use the `ctx` they may not know to update to use the `ctx` created from `StartSpan`. I feel like that would be an easy mistake to make (read: I would probably do that haha)

Reply:
Just looking at this, I agree, it seems weird that the linter would push you towards the more-likely-to-go-wrong version, because to avoid assigning a new value to `ctx`, you now have a stale value in `ctx`. There's no way to say "Don't use *this* if you need a context" in code, except assigning over the top.

Reply:
Had a silly thought to add a context to `oc.SetSpanStatus()` which (I thought) was called as defer for all cases. ... except for this one of course 😂. Doing so would satisfy the linter, and make it somewhat "natural" to not discard the returned context

i.e., for most, it would then look like

```go
ctx, span := trace.StartSpan(ctx, "DiagPid")
defer span.End()
defer func() { oc.SetSpanStatus(ctx, span, err) }()
```

Reply:
I'm not sure how I feel about doing that. I feel like at that point we should just have a function that noops with the ctx variable so we aren't adding unnecessary parameters when we don't need to be. Thoughts?

Reply:
I'd say that `span.End` should "consume" ctx back, like `span.End(ctx)`. Or `ctx.End(span)` if it is valid to have span without ctx. Or maybe span should be a field of ctx, so you say just `ctx.End`. What was started, should be ended. This way you don't need to suppress warnings, `ctx` is ready to be used not hidden as `_`, things are perfectly symmetrical and universe lives in peace. Anyway, my point is that returning a tuple of strongly-connected things could instead be expressed as returning a single thing that contains a second thing.

----

To be honest, I have no idea wtf is neither ctx nor span.

Reply:
Is ctx that is returned from StartSpan the same ctx that was passed to StartSpan? If yes, why it returns it at all?

Reply:
Another thought: what about `ctx.StartSpan` + `defer ctx.EndSpan` with a stack of spans inside ctx?

Reply:
`StartSpan` returns the context that the span is now running in, see [here](https://pkg.go.dev/go.opencensus.io@v0.23.0/trace#pkg-overview)

Reply:
we can't add new functions on the context struct type since it's defined in a different imported package that we don't own.

Reply:
If we keep the `nolint` directives in, and then later change the code such that they are no longer needed (e.g. start using `ctx`), will the linter tell us "hey, this directive isn't doing anything"?

Reply:
Not from what I've seen

Reply:
I don't have a strong preference here. I would probably go with `_, span := ...`, and rely on re-creating the `ctx` variable later if we end up needing it, but leaving the `nolint` directives in seems okay to me as well.

Reply:
I think I'm going to keep the `nolint` directives in for now and if it becomes annoying we can take them out.